docs/design/code_and_name_organization/README.md

```diff
@@ -671,6 +672,66 @@ namespace Shapes;
 struct Shapes.Square { ... };
 ```
 
+#### Declaring namespace members
+
+Namespace members may only be declared in the same name scope which was used to
+declare the namespace. For example:
+
+```carbon
+namespace NS;
+
+// ✅ Allowed: declaration is in file scope, which also declared `NS`.
+class NS.ClassT {
+  // ❌ Error: A class body has its own name scope.
+  var NS.a: i32 = 0;
+}
+
+fn Function() {
+  // ❌ Error: A function body has its own name scope.
+  var NS.b: i32 = 1;
+
+  namespace FunctionNS;
+  // ✅ Allowed: Both `FunctionNS` and `FunctionNS.c` are defined in the same
+  // name scope.
+  var FunctionNS.c: i32 = 2;
+
+  if (true) {
+    // ❌ Error: This is a separate name scope from `FunctionNS`.
+    var FunctionNS.d: i32 = 3;
+  }
+}
+
+// ✅ Allowed: declaration is in file scope, which also declared `NS`.
+namespace NS.MemberNS;
+
+// ✅ Allowed: declaration is in file scope, which also declared `NS.MemberNS`.
+class NS.MemberNS.MemberClassT {}
+```
+
+When there are multiple declared names in binding patterns, particularly `var`
+and `let`, all names must be in the same namespace. For example:
+
+```carbon
+namespace NS;
+
+// ✅ Allowed: `a` and `b` use the default namespace.
+var (a: i32, b: i32) = (1, 2);
+
+// ✅ Allowed: `c` and `d` are in the same namespace.
+var (NS.c: i32, NS.d: i32) = (3, 4);
+
+// ❌ Error: `e` and `f` are not in the same namespace.
+var (e: i32, NS.f: i32) = (5, 6);
+```
+
+This restriction only applies to binding patterns, not more general uses of
+patterns.
+
+```
+// ✅ Allowed: not a binding pattern.
+case (NS.namespace_var, local_var) => ...;
+```
```

Comment:
I didn't find this example helped my understanding, because it doesn't declare any names. Is there a case where it wouldn't be equivalent to say "all names declared by a pattern must be in the same namespace"?

Reply:
Removing the example, but I do think the clarification on binding patterns is important.

Reply:
If I'm understanding correctly, what you're distinguishing between here isn't binding patterns versus other kinds of patterns, it's names being declared in patterns versus names being used in patterns.

Eg, `NS1.x: NS2.y` should be OK, despite being a binding pattern and mentioning two different namespaces. And if we had a hypothetical `name@pattern` syntax to give a name to a thing matched by another pattern (eg, [as-patterns in Haskell](https://www.haskell.org/tutorial/patterns.html), corresponding [future work](https://github.com/carbon-language/carbon-lang/blob/trunk/proposals/p2188.md#as-patterns) for Carbon), then `(NS1.x@(_: auto, _: auto), NS2.y: i32)` would not be OK because it declares two names in different namespaces, even though one of them isn't using a binding pattern.

Reply:
> If I'm understanding correctly, what you're distinguishing between here isn't binding patterns versus other kinds of patterns, it's names being declared in patterns versus names being used in patterns.

Personally I would say when declaring names in binding patterns.

> `(NS1.x@(_: auto, _: auto), NS2.y: i32)`

Do I need to address `as` patterns? I'm not clear where the names introduced on the right would exist; the lack of a design makes it hard to respond to.

Reply:
What you have now looks fine to me. Saying that this is specific to binding patterns seems reasonable, given that they're the only way to declare names in a pattern currently. We can generalize this rule if and when we add as-patterns.